utils/rate.go

```diff
@@ -0,0 +1,97 @@
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+//
+// SOURCE: https://github.com/uber-go/ratelimit/blob/main/limiter_mutexbased.go
+// EDIT: slight modification to allow setting rate limit on the fly
+// SCOPE: LeakyBucket
+package utils
+
+import (
+	"sync"
+	"time"
+
+	"go.uber.org/atomic"
+)
+
+type LeakyBucket struct {
+	mutex      sync.Mutex
+	last       time.Time
+	sleepFor   time.Duration
+	perRequest atomic.Duration
+	maxSlack   time.Duration
+	clock      Clock
+}
+
+func NewLeakyBucket(rateLimit int, slack time.Duration, clock Clock) *LeakyBucket {
+	var lb LeakyBucket
+	lb.SetRateLimit(rateLimit)
+	lb.maxSlack = -1 * time.Duration(slack) * lb.perRequest.Load()
+	lb.clock = clock
+	return &lb
+}
+
+// SetRateLimit sets the underlying rate limit.
+// The setting may not be applied immediately.
+//
+// SetRateLimit is THREAD SAFE and NON-BLOCKING.
+func (lb *LeakyBucket) SetRateLimit(rateLimit int) {
+	lb.perRequest.Store(time.Second / time.Duration(rateLimit))
+}
+
+// Take blocks to ensure that the time spent between multiple Take calls
+// is on average time.Second/rate.
+//
+// Take is THREAD SAFE and BLOCKING.
+func (lb *LeakyBucket) Take() time.Time {
+	lb.mutex.Lock()
+	defer lb.mutex.Unlock()
+
+	now := lb.clock.Now()
+
+	// If this is our first request, then we allow it.
+	if lb.last.IsZero() {
+		lb.last = now
+		return lb.last
+	}
+
+	// sleepFor calculates how much time we should sleep based on
+	// the perRequest budget and how long the last request took.
+	// Since the request may take longer than the budget, this number
+	// can get negative, and is summed across requests.
+	lb.sleepFor += lb.perRequest.Load() - now.Sub(lb.last)
+
+	// We shouldn't allow sleepFor to get too negative, since it would mean that
+	// a service that slowed down a lot for a short period of time would get
+	// a much higher RPS following that.
+	if lb.sleepFor < lb.maxSlack {
+		lb.sleepFor = lb.maxSlack
+	}
+
+	// If sleepFor is positive, then we should sleep now.
+	if lb.sleepFor > 0 {
+		lb.clock.Sleep(lb.sleepFor)
```

Comment:
did you determine that sleeping with lock held is fine?

Reply:
If it is fine, a comment here explaining why it is okay to sleep with lock held. 

From experience, three months down the line if somebody is debugging some hairy issue, this (i. e. sleep with lock held) will pop up as somewhat of a mystery and if this is contributing to some performance issue or some bad behaviour or some bug. 

You have mentioned `BLOCKING`, but an explicit ack that there is a sleep with lock held and it is by design would give more confidence in somebody reading this three months later :-)

Reply:
Synced offline. For now, this is good enough for performance and safety. We will revisit this if needed.